Allow hostvars delegation

bcoca opened the pull request:
allow delegation to use hostvars in the values for connection related properties

fixes #70334 
fixes #22737  incorrect expectation, delegated host should have access to its own vars, not inventory_hostname ones, this PR allows access is wanted via `hostvars[inventory_hostname]`.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
vars

bcoca committed:
ensure hostvars are available on delegation

  also inventory_hostname must point to current host and not delegated one
  fix get_connection since it was still mixing original host vars and delegated ones

bmillemathias-1a commented on test/integration/targets/delegate_to/has_hostvars.yml:
```diff
@@ -0,0 +1,42 @@
+          - badlogin is not failed
+          - "'undefined' not in badlogin['msg']"
+
+    - name: delegate but try to use ivnentory_hostname data directly
```
typo: ivnentory_hostname instead of inventory_hostname

bmillemathias-1a commented on test/integration/targets/delegate_to/has_hostvars.yml:
```diff
@@ -0,0 +1,42 @@
+- name: ensure delegated host has hostvars avaiable for resolving connection
```
typo: avaiable -> available

bmillemathias-1a commented on test/integration/targets/delegate_to/connection_plugins/fakelocal.py:
```diff
@@ -0,0 +1,76 @@
+    description:
+        - This connection plugin just verifies parameters passed in
+    author: ansible (@core)
+    version_added: histerical
```
histerical ? hysterical ? historical ?

bcoca replied:
it was a compromize tween both

s-hertel commented on lib/ansible/executor/task_executor.py:
```diff
@@ -858,15 +855,14 @@ def _get_connection(self, variables, templar):
         return connection
 
-    def _get_persistent_connection_options(self, connection, variables, templar):
-        final_vars = combine_vars(variables, variables.get('ansible_delegated_vars', dict()).get(self._task.delegate_to, dict()))
```
This is a change so it could use a porting note, but looks like it's doing the right thing now.

bcoca replied:
yep, missed this in first PR of series that fixes delegation vars .. this was still doing the old 'combining all vars' so you would get from inventory_hostname if the delegated host was not overriding.

tempted to say i should add porting note for the original PR and ref the whole list of them up to this one.

bcoca committed:
also return connection vars for delegation

 add test to ensure we have expected usage when directly assigning for non delegated host

bcoca committed:
clog update

ci_complete

bcoca committed:
also return updated var info for non delegated

ci_complete

bcoca committed:
comment w/o whitespace

bcoca committed:
no passwords

bcoca committed:
spacer!t

jborean93 commented:
Backport PR to stable-2.10 https://github.com/ansible/ansible/pull/70855.
